packages/flutter_tools/lib/src/resident_web_runner.dart

```diff
@@ -130,76 +281,287 @@ class ResidentWebRunner extends ResidentRunner {
       printError(message);
       return 1;
     }
-    // Start the web compiler and build the assets.
-    await webCompilationProxy.initialize(
-      projectDirectory: FlutterProject.current().directory,
-      targets: <String>[target],
+    final Configuration configuration = Configuration(
+      autoRun: true,
+      debug: true,
+      hostname: 'localhost',
+      reload: ReloadConfiguration.hotRestart,
     );
-    _lastCompiled = DateTime.now();
-    final AssetBundle assetBundle = AssetBundleFactory.instance.createBundle();
-    final int build = await assetBundle.build();
-    if (build != 0) {
-      throwToolExit('Error: Failed to build asset bundle');
-    }
-    await writeBundle(fs.directory(getAssetBuildDirectory()), assetBundle.entries);
+    final Map<String, int> targetPorts = <String, int>{
+      'web': await os.findFreePort()
+    };
 
-    // Step 2: Start an HTTP server
-    _server = WebAssetServer(flutterProject, target, ipv6);
-    await _server.initialize();
+    /// Start the build daemon and run an initial build.
+    final List<String> buildOptions = <String>[];
+    final String workingDirectory = fs.currentDirectory.path;
+    _client = await _startBuildDaemon(workingDirectory, buildOptions);
+    _client.startBuild();
 
-    // Step 3: Spawn an instance of Chrome and direct it to the created server.
-    final String url = 'http://localhost:${_server.port}';
-    final Chrome chrome = await chromeLauncher.launch(url);
-    final ChromeTab chromeTab = await chrome.chromeConnection.getTab((ChromeTab chromeTab) {
-      return chromeTab.url.contains(url); // we don't care about trailing slashes or #
-    });
-    _connection = await chromeTab.connect();
-    _connection.onClose.listen((WipConnection connection) {
-      exit();
+    // Listen to build results to log error messages.
+    _buildResults = _client.buildResults.listen((BuildResults data) {
+      if (data.results.any((BuildResult result) =>
+          result.status == BuildStatus.failed ||
+          result.status == BuildStatus.succeeded)) {
+        printTrace(data.results.toString());
+      }
     });
 
-    // We don't support the debugging proxy yet.
+    // Register the current build targets.
+    _registerBuildTargets(_client, configuration, targetPorts);
+
+    // Start dev tools.
+    _devTools = await _startDevTools(configuration);
+
+    // Start the web dev server.
+    final int assetPort = daemonPort(workingDirectory);
+    _webdevServer = await WebDevServer.start(ServerOptions(
```

Comment:
We really only need the webdev server to get the port for chrome and to get the appDebugServices below. We could probably encapsulate this into a single class which took a configuration and returned the AppDebugServices

cc @grouma